sel/src/Sel/SecretKey/AuthenticatedEncryption.hs

```diff
@@ -0,0 +1,388 @@
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    cryptoSecretboxKeygen ptr
+  pure $ SecretKey fPtr
+
+-- | Create a 'SecretKey' from a 'StrictByteString' that you have obtained on your own,
+-- usually from the network or disk.
+--
+-- @since 0.0.1.0
+secretKeyFromByteString :: StrictByteString -> SecretKey
+secretKeyFromByteString bytestring = unsafeDupablePerformIO $
+  BS.unsafeUseAsCStringLen bytestring $ \(outsideSecretKeyPtr, _) -> do
+    secretKeyForeignPtr <- BS.mallocByteString @CChar (fromIntegral cryptoSecretboxKeyBytes)
+    Foreign.withForeignPtr secretKeyForeignPtr $ \secretKeyPtr ->
+      Foreign.copyArray outsideSecretKeyPtr secretKeyPtr (fromIntegral cryptoSecretboxKeyBytes)
+    pure $ SecretKey (Foreign.castForeignPtr @CChar @CUChar secretKeyForeignPtr)
+
+-- | Convert a 'SecretKey' to a strict, hexadecimal-encoded 'Text'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+secretKeyToHexText :: SecretKey -> Text
+secretKeyToHexText = Base16.encodeBase16 . secretKeyToBinary
+
+-- | Convert a 'SecretKey' to a strict, hexadecimal-encoded 'StrictByteString'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+secretKeyToHexByteString :: SecretKey -> StrictByteString
+secretKeyToHexByteString = Base16.encodeBase16' . secretKeyToBinary
+
+-- | Convert a 'SecretKey' to a strict binary 'StrictByteString'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+secretKeyToBinary :: SecretKey -> StrictByteString
+secretKeyToBinary (SecretKey secretKeyForeignPtr) =
+  BS.fromForeignPtr0
+    (Foreign.castForeignPtr @CUChar @Word8 secretKeyForeignPtr)
+    (fromIntegral @CSize @Int cryptoSecretboxKeyBytes)
+
+-- | A random number that must only be used once per exchanged message.
+-- It does not have to be confidential.
+-- It is of size 'cryptoSecretboxNonceBytes'.
+--
+-- @since 0.0.1.0
+newtype Nonce = Nonce (ForeignPtr CUChar)
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq Nonce where
+  (Nonce hk1) == (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretboxKeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord Nonce where
+  compare (Nonce hk1) (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretboxKeyBytes
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Display Nonce where
+  displayBuilder = Builder.fromText . nonceToHexText
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Show Nonce where
+  show = BS.unpackChars . nonceToHexByteString
+
+-- | Generate a new random nonce.
+-- Only use it once per exchanged message.
+--
+-- Do not use this outside of hash creation!
+newNonce :: IO Nonce
+newNonce = do
+  (fPtr :: ForeignPtr CUChar) <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretboxNonceBytes)
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    randombytesBuf (Foreign.castPtr @CUChar @Word8 ptr) cryptoSecretboxNonceBytes
+  pure $ Nonce fPtr
```

Comment:
Why would you even need this? Nonces can only show up in two ways that make sense:

* Read from the outside world; or
* During the process of encryption.

This helps neither, and acts as a very large footgun.

Reply:
It's not exposed, I just need it in the `encrypt` function.

Reply:
Ah, OK. Then why does it have Haddocks?

Reply:
Because it's so damn convenient :) 
![image](https://github.com/haskell-cryptography/libsodium-bindings/assets/29253044/95bfd130-4244-40bc-ad1b-805c47b208fc)

Reply:
I genuinely don't believe 'IDE support' is sufficient reason for Haddocks on a private helper. However, I'm willing to concede that one. I just find it personally confusing.